Improve invalid let expression handling

matthewjasper opened the pull request:
- Move all of the checks for valid let expression positions to parsing. 
- Add a field to ExprKind::Let in AST/HIR to mark whether it's in a valid location.
- Suppress some later errors and MIR construction for invalid let expressions.
- Fix a (drop) scope issue that was also responsible for #104172.

Fixes #104172
Fixes #104868

matthewjasper committed:
Don't create drop scopes after item statements

These scopes would not exist in MIR and can cause ICEs with invalid uses
of let expressions.

matthewjasper committed:
Move let expression checking to parsing

There was an incomplete version of the check in parsing and a second
version in AST validation. This meant that some, but not all, invalid
uses were allowed inside macros/disabled cfgs. It also means that later
passes have a hard time knowing when the let expression is in a valid
location, sometimes causing ICEs.

- Add a field to ExprKind::Let in AST/HIR to mark whether it's in a
  valid location.
- Suppress later errors and MIR construction for invalid let
  expressions.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6123543513/job/16621651300) [(plain)](https://github.com/rust-lang/rust/commit/9c8c863f4c8ff221b1f092b058189081f940fe72/checks/16621651300/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking tracing-error v0.2.0
error[E0023]: this pattern has 3 fields, but the corresponding tuple variant has 4 fields
    --> src/tools/clippy/clippy_utils/src/ast_utils.rs:169:14
     |
169  |         (Let(lp, le, _), Let(rp, re, _)) => eq_pat(lp, rp) && eq_expr(le, re),
     |              ^^  ^^  ^ expected 4 fields, found 3
    ::: /checkout/compiler/rustc_ast/src/ast.rs:1429:9
     |
     |
1429 |     Let(P<Pat>, P<Expr>, Span, Option<ErrorGuaranteed>),
     |
help: use `_` to explicitly ignore each field
     |
     |
169  |         (Let(lp, le, _, _), Let(rp, re, _)) => eq_pat(lp, rp) && eq_expr(le, re),

error[E0023]: this pattern has 3 fields, but the corresponding tuple variant has 4 fields
    --> src/tools/clippy/clippy_utils/src/ast_utils.rs:169:30
     |
     |
169  |         (Let(lp, le, _), Let(rp, re, _)) => eq_pat(lp, rp) && eq_expr(le, re),
     |                              ^^  ^^  ^ expected 4 fields, found 3
    ::: /checkout/compiler/rustc_ast/src/ast.rs:1429:9
     |
     |
1429 |     Let(P<Pat>, P<Expr>, Span, Option<ErrorGuaranteed>),
     |
help: use `_` to explicitly ignore each field
     |
     |
169  |         (Let(lp, le, _), Let(rp, re, _, _)) => eq_pat(lp, rp) && eq_expr(le, re),

   Compiling futures-macro v0.3.28
For more information about this error, try `rustc --explain E0023`.
error: could not compile `clippy_utils` (lib) due to 2 previous errors
```

</details>

matthewjasper committed:
Update tools and fulldeps tests

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6124022704/job/16623179812) [(plain)](https://github.com/rust-lang/rust/commit/7964b11bdf02a5828d63239e66167bfe4cd3e7d4/checks/16623179812/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
--- stderr -------------------------------
error[E0061]: this enum variant takes 4 arguments but 3 arguments were supplied
##[error]  --> /checkout/tests/ui-fulldeps/pprust-expr-roundtrip.rs:177:50
   |
LL |                 iter_exprs(depth - 1, &mut |e| g(ExprKind::Let(pat.clone(), e, DUMMY_SP)))
   |                                                  ^^^^^^^^^^^^^-------------------------- an argument of type `Option<ErrorGuaranteed>` is missing
note: tuple variant defined here
  --> /rustc/FAKE_PREFIX/compiler/rustc_ast/src/ast.rs:1429:5
help: provide the argument
   |
   |
LL |                 iter_exprs(depth - 1, &mut |e| g(ExprKind::Let(pat.clone(), e, DUMMY_SP, /* Option<ErrorGuaranteed> */)))

error: aborting due to previous error

For more information about this error, try `rustc --explain E0061`.
```

</details>

matthewjasper committed:
Reduce double errors for invalid let expressions

Previously some invalid let expressions would result in both a feature
error and a parsing error. Avoid this and ensure that we only emit the
parsing error when this happens.

matthewjasper commented on compiler/rustc_parse/src/parser/expr.rs:
```diff
@@ -1345,110 +1348,113 @@ impl<'a> Parser<'a> {
+                    this.recover_incorrect_await_syntax(lo, this.prev_token.span)
                 } else {
-                    self.parse_expr_closure()
+                    this.parse_expr_lit()
```
The diff isn't showing the changes here very well (maybe due to the extra indent). There isn't actually a change in the methods being called here.

matthewjasper committed:
Address review comments

- Add doc comment to new type
- Restore "only supported directly in conditions of `if` and `while` expressions" note
- Rename variant with clearer name

b-naber commented:
@bors r+

bors commented:
:pushpin: Commit e324a59eb6ba1a7883bf23ff42d425ca96960f2a has been approved by `b-naber`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=b-naber e324a59eb6ba1a7883bf23ff42d425ca96960f2a -->
<!-- homu: {"type":"Approved","sha":"e324a59eb6ba1a7883bf23ff42d425ca96960f2a","approver":"b-naber","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e324a59eb6ba1a7883bf23ff42d425ca96960f2a with merge 9d21092b8548aa80c411c7c18d4266c092352b33...
<!-- homu: {"type":"BuildStarted","head_sha":"e324a59eb6ba1a7883bf23ff42d425ca96960f2a","merge_sha":"9d21092b8548aa80c411c7c18d4266c092352b33"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6180356863/job/16778347776)
Approved by: b-naber
Pushing 9d21092b8548aa80c411c7c18d4266c092352b33 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"b-naber","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6180356863/job/16778347776"},"merge_sha":"9d21092b8548aa80c411c7c18d4266c092352b33"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

rust-timer commented:
Finished benchmarking commit (9d21092b8548aa80c411c7c18d4266c092352b33): [comparison URL](https://perf.rust-lang.org/compare.html?start=8142a319ed5c1d1f96e5a1881a6546e463b77c8f&end=9d21092b8548aa80c411c7c18d4266c092352b33&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=8142a319ed5c1d1f96e5a1881a6546e463b77c8f&end=9d21092b8548aa80c411c7c18d4266c092352b33&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.3%, 0.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8142a319ed5c1d1f96e5a1881a6546e463b77c8f&end=9d21092b8548aa80c411c7c18d4266c092352b33&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=8142a319ed5c1d1f96e5a1881a6546e463b77c8f&end=9d21092b8548aa80c411c7c18d4266c092352b33&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.6% | [0.6%, 0.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.6% | [0.6%, 0.6%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=8142a319ed5c1d1f96e5a1881a6546e463b77c8f&end=9d21092b8548aa80c411c7c18d4266c092352b33&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.556s -> 632.444s (0.14%)
**Artifact size**: 317.95 MiB -> 317.98 MiB (0.01%)

b-naber commented:
@bors retry

bors commented:
:hourglass: Testing commit e324a59eb6ba1a7883bf23ff42d425ca96960f2a with merge dac91a82e146d89fefe94dcbb53cd6c2662e354f...
<!-- homu: {"type":"BuildStarted","head_sha":"e324a59eb6ba1a7883bf23ff42d425ca96960f2a","merge_sha":"dac91a82e146d89fefe94dcbb53cd6c2662e354f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6190250242/job/16809023338)
Approved by: b-naber
Pushing dac91a82e146d89fefe94dcbb53cd6c2662e354f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"b-naber","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6190250242/job/16809023338"},"merge_sha":"dac91a82e146d89fefe94dcbb53cd6c2662e354f"} -->

rust-timer commented:
Finished benchmarking commit (dac91a82e146d89fefe94dcbb53cd6c2662e354f): [comparison URL](https://perf.rust-lang.org/compare.html?start=ccf817b9bbe449204a227430d0a84a4f1d1798cc&end=dac91a82e146d89fefe94dcbb53cd6c2662e354f&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ccf817b9bbe449204a227430d0a84a4f1d1798cc&end=dac91a82e146d89fefe94dcbb53cd6c2662e354f&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.2%, 0.3%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ccf817b9bbe449204a227430d0a84a4f1d1798cc&end=dac91a82e146d89fefe94dcbb53cd6c2662e354f&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.2% | [4.2%, 4.2%] | 1     |
| Regressions ❌ <br /> (secondary)  | 1.0% | [0.7%, 1.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.2% | [4.2%, 4.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ccf817b9bbe449204a227430d0a84a4f1d1798cc&end=dac91a82e146d89fefe94dcbb53cd6c2662e354f&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=ccf817b9bbe449204a227430d0a84a4f1d1798cc&end=dac91a82e146d89fefe94dcbb53cd6c2662e354f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.307s -> 631.195s (-0.02%)
**Artifact size**: 317.98 MiB -> 318.01 MiB (0.01%)
